### PR TITLE
CastXML integration

### PR DESCRIPTION
**Pull request by goldhoorn**
I started to try to integrate castxml, but it seems that this is NOT a drop-in-replacement. 

I currently generate all outputs from gccxml and castxml (castxml is active)
the calls to castxml wen trught, but the gnerated registry is empty.
Someone has to dig into this why the castxml output is not processed correctly.

This is work for either
@doudou, or (Team 8 responsibility) @jmachowinski @saarnold or @jhidalgocarrio

### CODE REVIEW

**Comment by doudou** [comment]
Hey, thanks.

You forgot to change the castxml handler to use .hpp files ... (I say "forgot" as you did it in the preprocess method)

**Comment by doudou** [comment]
Still not working with that change, though.

**Comment by goldhoorn** [comment]
This is not needed because THW file is not processed later. Only THW string is returned .

However semms a bit of work. I can provide tomorrow the preprocessed files.

**Comment by goldhoorn** [comment]
THW = the (Handy german correction argh)

**Comment by doudou** [comment]
I've brought the stuff up to a point where it looks promising. Basically, the "major" change is that the 'demangled' field has disappeared. The thing is, it was not always there and so using it was making the code that much more complicated.

Code's not clean either. Let's get it to work first.

**Comment by goldhoorn** [comment]
Please keep ind mind that this change is also needed (sorry i forgot to push yesterday):
https://github.com/orocos-toolchain/orogen/pull/64

**Comment by doudou** [comment]
> Please keep ind mind that this change is also needed (sorry i forgot to push yesterday)

Actually, it was not (but I merged it anyways, forgetting about that)

The performance optimization of putting all includes in a single temporary file fails with castxml (the type-to-file mappings are not reported properly when doing that). See my comment above.

**Comment by goldhoorn** [comment]
The all in one file causes also this side-problem (or at least would solve it)
https://github.com/orocos-toolchain/orogen/issues/50

**Comment by goldhoorn** [comment]
Ok i spend a bit work in here, the current state is that the importer now roughly works. 
I hit a few bugs in typelib (specially the return string <-> nil) problem.

However the current state is that the base/orogen/std could be generated successfull (as far i see).
The problem with base/orogen/types is now definitily the same one as in https://github.com/orocos-toolchain/orogen/issues/50 it seems the headers are reported in a different order by castxml.

However, the orogen bug must be solved before any further development can be done here.
The bug occus here even on a (not-export-all-typekits) setup.

**Comment by goldhoorn** [comment]
@marvin2k (cc)

**Comment by goldhoorn** [comment]
For the sake of clarity: i will not work on this include bug if this header resolution is fixed i might have time to check castxml itself and help for a clear integration. But this special bug needs to be solved by someone else.

File: bindings/ruby/lib/typelib/gccxml.rb [hunk 1]
```diff
@@ -695,6 +647,7 @@ def self.parse_cxx_documentation_before(lines, line)
         def load(required_files, xml)
             @registry = Typelib::Registry.new
             Typelib::Registry.add_standard_cxx_types(registry)
+            registry.alias('/std/basic_string</char>', '/std/string')
```
Comment (marvin2k) [inline bindings/ruby/lib/typelib/gccxml.rb]: Shouldn't this be moved into `add_standard_cxx_types`? It left  my memory why this manual alias is not needed for the gccxml-case...

File: bindings/ruby/lib/typelib/gccxml.rb [hunk 2]
```diff
@@ -760,15 +713,74 @@ class << self
...
+            end
+            return "gccxml"
+
+        end
+        # Runs castxml on the provided file and with the given options, and
+        # return the Nokogiri::XML object representing the result
+        #
+        # Raises RuntimeError if casrxml failed to run
+        def self.castxml(file, options)
+            cmdline = [use_castxml, *castxml_default_options, "--castxml-gccxml", '-x', 'c++', '-std=c++11']
```
Comment (marvin2k) [inline bindings/ruby/lib/typelib/gccxml.rb]: this enforced "c++11" needs to be cleaned up

Comment (goldhoorn) [inline bindings/ruby/lib/typelib/gccxml.rb]: On 06.10.2015 10:12, marvin2k wrote:
>
> In bindings/ruby/lib/typelib/gccxml.rb 
> <https://github.com/orocos-toolchain/typelib/pull/66#discussion_r41235579>:
>
> > +            return "gccxml"
> > +        end
> > +
> > +        def self.use_castxml
> > +            if !`which castxml`.empty?
> > +                return "castxml"
> > +            end
> > +            return "gccxml"
> > +
> > +        end
> > +        # Runs castxml on the provided file and with the given options, and
> > +        # return the Nokogiri::XML object representing the result
> > +        #
> > +        # Raises RuntimeError if casrxml failed to run
> > +        def self.castxml(file, options)
> > +            cmdline = [use_castxml, *castxml_default_options, "--castxml-gccxml", '-x', 'c++', '-std=c++11']
>
> this enforced "c++11" needs to be cleaned up
>
Should be checked/setted as a ENV (maybe then set by autoproj) depending 
weather we build c++11 in the future
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/orocos-toolchain/typelib/pull/66/files#r41235579>.
>


-- 
  Dipl.-Inf. Matthias Goldhoorn
  Space and Underwater Robotic

  Universität Bremen
  FB 3 - Mathematik und Informatik
  AG Robotik
  Robert-Hooke-Straße 1
  28359 Bremen, Germany
  
  Zentrale: +49 421 178 45-6611
  
  Besuchsadresse der Nebengeschäftstelle:
  Robert-Hooke-Straße 5
  28359 Bremen, Germany
  
  Tel.:    +49 421 178 45-4193
  Empfang: +49 421 178 45-6600
  Fax:     +49 421 178 45-4150
  E-Mail:  matthias.goldhoorn@informatik.uni-bremen.de

  Weitere Informationen: http://www.informatik.uni-bremen.de/robotik

Comment (marvin2k) [inline bindings/ruby/lib/typelib/gccxml.rb]: jo

**Comment by doudou** [comment]
FYI, I've cleaned up the patches, and -- much more importantly -- moved the clang test suite to the common parts and apply it to gccxml, clang and castxml. Currently gccxml/castxml pass most of them (one does not pass because it depends on Eigen ... @marvin2k could you remove the dependency ?). The other one I've not looked into it yet.

What is tested so far is the ability to define the types properly. Metadata is not tested, but it will (especially given how important it is to the orogen side)

I'll from now on have a policy that every single bugfix in the C++ importer should have a corresponding test added.

**Comment by goldhoorn** [comment]
@doudou: do you have a general concept (in mind) how we handle the cxx11 abi's?
Is really ignoring (as here: https://github.com/orocos-toolchain/typelib/commit/64a309d118b54367fdfbd373cf18f1fe653cfa42) the way to go?

**Comment by marvin2k** [comment]
Hm, I added the eigen dependency to check the opaque mechanism. Removed them now. I don't know how to run the tests now, so testing this commit is left for the curious among us ;-)

Thanks for attacking this castxml problem, this removes the need for a lot of long-time work on the clang importer!

**Comment by doudou** [comment]
> Ah, I did not edit the generated templates.tlb as it will be generated automatically by running the testsuite?

Actually not, the tlb files are used as expected result and compared with the importer result.

**Comment by marvin2k** [comment]
So how are the expected results generated then? By hand? In the cmake-based testing the tlb's were generated in the build-dir and had to be copied manually into the src-folder.

**Comment by doudou** [comment]
> So how are the expected results generated then? By hand? In the cmake-based testing the tlb's were generated in the build-dir and had to be copied manually into the src-folder.

That's what I had in mind to add new "expected" file. Generate it, check it by hand to make sure it matches and then add it to the test folder.

**Comment by marvin2k** [comment]
The CMakeList with the target is "gone". I thought this is somehow replaced by minitest?

**Comment by doudou** [comment]
I'm trying to iron out a process here to 
 - easily run a ruby test (currently easy to run the whole test suite, but a pain to run a single test)
 - save the generated tlb files in the build directory

I'll come back to you

**Comment by marvin2k** [comment]
in shell-speak:
```bash
./build/tools/typelib-clang-tlb-importer/typelib-clang-tlb-importer test/ruby/cxx_import_tests/templates.hh -opaquePath=test/ruby/cxx_import_tests/templates.opaques -tlbSavePath=templates.tlb
```

**Comment by doudou** [comment]
> in shell-speak:

This assumes that you have a clang importer.

**Comment by doudou** [comment]
Rebased and updated things. The castxml and gccxml importers now pass the test suite

One major problem is that castxml does not output the offset of base classes (meaning, it will break for multiple inheritance). I've added a ticket to castxml about that (https://github.com/CastXML/CastXML/issues/34)

**Comment by doudou** [comment]
To run the tests, there's a runner script in build/test/ruby. Just cd in there and run

     ./runner
     ./runner test_cxx_gccxml.rb
     ./runner test_cxx_gccxml.rb -n /test_name_filter/

To generate the tlbs, run

     ./cxx_tlbgen name_of_test_file.hh

**Comment by hauptmech** [comment]
I think I need castxml integrated to build rock with gcc5. How's this going?

**Comment by doudou** [comment]
the status is pretty good, but there are still some issues with some orogen projects in Rock. Castxml has re-integrated class offsets in its gccxml output format, so that is also supported now.

I don't have time to work on it at all right now.

**Comment by doudou** [comment]
The castxml integration can now handle buildconf-all. I remove the "do not merge" label.

Corresponding autoproj integration: https://github.com/orocos-toolchain/autoproj/pull/16 and https://github.com/rock-core/package_set/pull/74

**Comment by doudou** [comment]
Given the 98% lack of review of PRs (important or otherwise) from anyone to rock-core in the last year or so, I merge this.

**Pull request merged**